rinja_derive/src/input.rs

```diff
@@ -422,6 +433,177 @@ impl TemplateArgs {
+fn source_from_docs(
+    name: &syn::Ident,
+    value: &syn::ExprLit,
+    dest: &mut Option<(Source, Option<Span>)>,
+    ast: &syn::DeriveInput,
+) -> Result<(), CompileError> {
+    match &value.lit {
+        syn::Lit::Bool(syn::LitBool { value, .. }) => {
+            if !value {
+                return Ok(());
+            }
+        }
+        lit => {
+            return Err(CompileError::no_file_info(
+                "argument `in_doc` expects as boolean value",
+                Some(lit.span()),
+            ));
+        }
+    };
+    #[cfg(not(feature = "code-in-doc"))]
+    {
+        let _ = (name, dest, ast);
+        Err(CompileError::no_file_info(
+            "enable feature `code-in-doc` to use `in_doc` argument",
+            Some(name.span()),
+        ))
+    }
+    #[cfg(feature = "code-in-doc")]
+    {
+        ensure_source_once(name, dest)?;
+        let (span, source) = collect_comment_blocks(name, ast)?;
+        let source = strip_common_ws_prefix(source);
+        let source = collect_rinja_code_blocks(name, ast, source)?;
+        *dest = Some((source, span));
+        Ok(())
+    }
+}
+
+#[cfg(feature = "code-in-doc")]
+fn collect_comment_blocks(
+    name: &syn::Ident,
+    ast: &syn::DeriveInput,
+) -> Result<(Option<Span>, String), CompileError> {
+    let mut span: Option<Span> = None;
+    let mut assign_span = |kv: &syn::MetaNameValue| {
+        // FIXME: uncomment once <https://github.com/rust-lang/rust/issues/54725> is stable
+        // let new_span = kv.path.span();
+        // span = Some(match span {
+        //     Some(cur_span) => cur_span.join(new_span).unwrap_or(cur_span),
+        //     None => new_span,
+        // });
+
+        if span.is_none() {
+            span = Some(kv.path.span());
+        }
+    };
+
+    let mut source = String::new();
+    for a in &ast.attrs {
+        // is a comment?
+        let syn::Meta::NameValue(kv) = &a.meta else {
+            continue;
+        };
+        if !kv.path.is_ident("doc") {
+            continue;
+        }
+
+        // is an understood comment, e.g. not `#[doc = inline_str(…)]`
+        let mut value = &kv.value;
+        let value = loop {
+            match value {
+                syn::Expr::Lit(lit) => break lit,
+                syn::Expr::Group(group) => value = &group.expr,
+                _ => continue,
+            }
+        };
+        let syn::Lit::Str(value) = &value.lit else {
+            continue;
+        };
+
+        assign_span(kv);
+        source.push_str(value.value().as_str());
+        source.push('\n');
+    }
+    if source.is_empty() {
+        return Err(no_rinja_code_block(name, ast));
+    }
+
+    Ok((span, source))
+}
+
+#[cfg(feature = "code-in-doc")]
+fn no_rinja_code_block(name: &syn::Ident, ast: &syn::DeriveInput) -> CompileError {
+    let kind = match &ast.data {
+        syn::Data::Struct(_) => "struct",
+        syn::Data::Enum(_) => "enum",
+        syn::Data::Union(_) => "union",
```

Comment:
So we do support all item types! :laughing:

Reply:
Hehe, I simply added cases for all variant of the enum. I actually don't think that we generate valid code for unions, because I think reading from unions is only allowed in an unsafe context.

Reply:
Funny. Might worth handling unions correctly at some point.